src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+        (event.events & libc::POLLHUP) != 0
+            // Socket has received FIN or called shutdown(SHUT_RD)
+            || (event.events & POLLRDHUP) != 0
+    }
+
+    pub fn is_write_closed(event: &Event) -> bool {
+        // Both halves of the socket have closed
+        (event.events & libc::POLLHUP) != 0
+            // Unix pipe write end has closed
+            || ((event.events & libc::POLLOUT) != 0 && (event.events & libc::POLLERR) != 0)
+            // The other side (read end) of a Unix pipe has closed.
+            || (event.events == libc::POLLERR)
+    }
+
+    pub fn is_priority(event: &Event) -> bool {
+        (event.events & libc::POLLPRI) != 0
+    }
+
+    pub fn is_aio(_: &Event) -> bool {
+        // Not supported in the kernel, only in libc.
+        false
+    }
+
+    pub fn is_lio(_: &Event) -> bool {
+        // Not supported.
+        false
+    }
+
+    pub fn debug_details(f: &mut fmt::Formatter<'_>, event: &Event) -> fmt::Result {
+        #[allow(clippy::trivially_copy_pass_by_ref)]
+        fn check_events(got: &libc::c_short, want: &libc::c_short) -> bool {
+            (*got & want) != 0
+        }
+        debug_detail!(
+            EventsDetails(libc::c_short),
+            check_events,
+            libc::POLLIN,
+            libc::POLLPRI,
+            libc::POLLOUT,
+            libc::POLLRDNORM,
+            libc::POLLRDBAND,
+            libc::POLLWRNORM,
+            libc::POLLWRBAND,
+            libc::POLLERR,
+            libc::POLLHUP,
+        );
+
+        // Can't reference fields in packed structures.
+        let e_u64 = event.token.0;
+        f.debug_struct("epoll_event")
+            .field("events", &EventsDetails(event.events))
+            .field("u64", &e_u64)
+            .finish()
+    }
+}
+
+cfg_io_source! {
+    use crate::Registry;
+
+    struct InternalState {
+        selector: Selector,
+        token: Token,
+        interests: Interest,
+        fd: RawFd,
+        shared_record: Arc<RegistrationRecord>,
+    }
+
+    impl Drop for InternalState {
+        fn drop(&mut self) {
+            if self.shared_record.is_registered() {
+                let _ = self.selector.deregister(self.fd);
+            }
+        }
+    }
+
+    pub(crate) struct IoSourceState {
+        inner: Option<Box<InternalState>>,
+    }
+
+    impl IoSourceState {
+        pub fn new() -> IoSourceState {
+            IoSourceState { inner: None }
+        }
+
+        pub fn do_io<T, F, R>(&self, f: F, io: &T) -> io::Result<R>
+        where
+        F: FnOnce(&T) -> io::Result<R>,
+        {
+            let result = f(io);
+
+            if let Err(err) = &result {
+            if err.kind() == io::ErrorKind::WouldBlock {
+                self.inner.as_ref().map_or(Ok(()), |state| {
+                    state
+                    .selector
+                    .reregister(state.fd, state.token, state.interests)
+                })?;
+            }
+        }
```

Comment:
No automatic formatting within macro :(
```suggestion
                if err.kind() == io::ErrorKind::WouldBlock {
                    self.inner.as_ref().map_or(Ok(()), |state| {
                        state
                        .selector
                        .reregister(state.fd, state.token, state.interests)
                    })?;
                }
            }
```